### PR TITLE
[mono][interp] Optimize box+unbox to no-op

### PR DESCRIPTION
**Pull request by EgorBo**
Common in generics, simple example:
```csharp
static int ToInt<T>(T t) => (int)(object)t; // can't cast directly T to VT
```
`MONO_VERBOSE_METHOD=ToInt`:
#### Before:
```
Runtime method: Repro:ToInt<int> (int) 0x7ffee273b6d0
Locals size 8, VT stack size: 0
Calculated stack size: 1, stated size: 8
IR_0000: ldloc.i4   0
IR_0002: box        0,0.  <--- expensive redundant boxing.
IR_0005: unbox      1
IR_0007: ldind.i4.check
IR_0008: ret
```
#### After
```
Runtime method: Repro:ToInt<int> (int) 0x7ffee77a06d0
Locals size 8, VT stack size: 0
Calculated stack size: 1, stated size: 8
IR_0000: ldloc.i4   0
IR_0002: ret
```

RyuJIT and Mono JIT both have this optimization.

### CODE REVIEW

File: src/mono/mono/mini/interp/transform.c [hunk 1]
```diff
@@ -4938,6 +4938,14 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 			klass = mini_get_class (method, token, generic_context);
 			CHECK_TYPELOAD (klass);
 
+			// Common in generic code:
+			// box T + unbox.any T -> nop
+			if (td->last_ins->opcode == MINT_BOX && (td->sp - 1)->klass == klass) {
+				interp_clear_ins (td, td->last_ins);
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:4944]: You also need to update the stack type of the top of the stack. Something like `SET_TYPE (td->sp - 1, stack_type [mt], klass);`

**Comment by BrzVlad** [comment]
Does this pattern show up also with valuetypes, which emit MINT_BOX_VT (as opposed to primitive types)

File: src/mono/mono/mini/interp/transform.c (hunk 1 above)
Comment (EgorBo) [reply src/mono/mono/mini/interp/transform.c:4944]: ah, I see, thanks!

**Comment by EgorBo** [comment]
> MINT_BOX_VT

Yeah it does! However, when I add it to my condition it fails with 
```
Program:ToDateTime<System.DateTime> (System.DateTime): CEE_RET: value type stack: 0 vs. 8
```
🤔

**Comment by BrzVlad** [comment]
@EgorBo You should include your handling of box_vt so I can see what is wrong. You are probably not updating the vtstack accordingly after removing instructions that push/pop it.

**Comment by EgorBo** [comment]
@BrzVlad can you please take a look at the latest commit, it seems it handles well BOX_VT now, e.g. for `ToDateTime` method IL is now:
```
Runtime method: Program:ToDateTime<System.DateTime> (System.DateTime) 0x7ffd189e4f00
Locals size 8, VT stack size: 8
Calculated stack size: 1, stated size: 8
IR_0000: ldarg.vt   0,8
IR_0004: ret.vt     8
```

**Review by BrzVlad — APPROVED** [review]
you could probably remove some duplicated code

**Comment by BrzVlad** [comment]
@EgorBo I forgot that you also need to check that the box and unbox are in the same basic block, otherwise you are not allowed to clear it. You would need to do a check for `td->is_bb_start`

**Pull request merged**